newstitles.py

```diff
@@ -0,0 +1,192 @@
+    """
+    Returns a list of all titles found as specified by the XPath `selector`, if any.
+
+    :param html: The HTML data to analyze
+    :param selector: The XPath selector of the titles
+    :return: A list of all titles found in the given HTML data using the given selector
+    """
+    try:
+        html_tree = html.fromstring(html_data)  # WARNING: Potentially heavy operation if input data is large enough
+        return [unicode(x).strip() for x in html_tree.xpath(selector)]
+    except Exception:
+        return []
+
+
+def create_frequency_table(titles):
+    """
+    Returns a sorted dict containing all tokens inside the given titles and their frequency among all titles.
+
+    To make the table smaller, it will remove those words with frequency = 1. Why? When we're doing a lookup in the
+    frequency table and some word is missing, we will know it's the first time we analyze it.
+
+    e.g. if `titles` = ['Trump goes to Japan', 'Trump offends Australian people'], the frequency table will look like:
+    {
+        'trump': 2,
+    }
+
+    :param html: The HTML data to analyze
+    :param selector: The XPath selector of the titles
+    :return: A list of all titles found in the given HTML data using the given selector
+    """
+    frequency_table = Counter()
+    for title in titles:
+        # Calculating the frequency of words in the current title and adding it to the main table
+        frequency_table.update(Counter(_get_all_words(title)))
+    # Removing words with frequency = 1 to make the table smaller
+    frequency_table = {k: v for k, v in frequency_table.items() if v > 1}
+    return frequency_table
+
+
+def get_titles_by_weight(titles, frequency_table=None, sort=False):
+    """
+    Returns a list of all titles and their weight as determined by the given frequency table.
+
+    :param titles: All titles to analyze
+    :param frequency_table: The frequency of all token's occurrences among all titles
+    :param sort: Whether to sort the final list of titles or not (in descending order)
+    :return: A list of all titles and their weight
+    """
+    if not frequency_table:
+        frequency_table = {}
+
+    weights_data = []  # A list of tuples of (weight, title)
+    for title in titles:
+        title_weight = _get_title_weight(title, frequency_table=frequency_table)
+        weights_data.append((title_weight, title))
+
+    if sort:
+        weights_data = sorted(weights_data, key=lambda x: x[0], reverse=True)
+    return weights_data
+
+
+def _get_all_words(text):
+    """
+    Internal method that uses regex to return all words found in a given text.
+
+    It doesn't include single chars found after regex operation runs (e.g. like when `it's` becomes `it s`).
+    All returned words will be in lowercase.
+
+    :param text: The text to find words in
+    :return: All words found in text
+    """
+    return [word.lower() for word in re.findall(r'\b([a-zA-Z]+)\b', text) if len(word) > 1]
```

Comment:
Should just split on whitespace, e.g `s.split()`, as `re.*` is slow/overkill and `.split()` is much faster

Reply:
I didn't think of this one, it could worth trying it